File: mvc1/_etc/Java/11727 2xn 타일링 2.java
import java.util.Scanner;

public class Main {
    public static void main(String args[]) {
      Scanner s = new Scanner(System.in);
      int n = s.nextInt();
      long[] dp = new long[1001];
      
      dp[0] = 0;
      dp[1] = 1;
      dp[2] = 3;
      dp[3] = 5;
      
      for(int i = 4; i < n+1; i++) {
          dp[i] = (dp[i-1] + 2*dp[i-2])%10007;
      }
      
      System.out.println(dp[n]);
      
    }
}

// http://boj.kr/db73c86a7c7f41a9a500769bd0829a63

File: Dynamic Programming/Kotlin/11727 2xn 타일링 2.kt
import java.util.Scanner
import kotlin.math.*

fun main(args: Array<String>) {
    val sc: Scanner = Scanner(System.`in`)
    val num = sc.nextInt()
    
    var dp = LongArray(1001, {0})
    
    dp[0] = 0
    dp[1] = 1
    dp[2] = 3
    dp[3] = 5
    
    for(i in 4..num) {
        dp[i] = (dp[i-1] + 2*dp[i-2]) % 10007
    }
    
    println("${dp[num]}")

}

// http://boj.kr/72f421fc4b9a45aaa66502b15c0a5f80
